src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -37,6 +38,40 @@ Aws::String Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &a
     }
 }
 
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+#ifdef USE_AWS_MEMORY_MANAGEMENT
+static const char* ZlibMemTag = "zlib";
+static size_t offset = sizeof(size_t); // to make space for size of the array
+//Define custom memory allocation for zlib
+// if fail to allocate, return Z_NULL
+void* aws_zalloc(void * /* opaque */, unsigned items, unsigned size)
+{
+    unsigned sizeToAllocate = items*size;
+    char* newMem = reinterpret_cast<char*>(Aws::Malloc(ZlibMemTag, sizeToAllocate + offset));
+    if (newMem != nullptr) {
+        std::size_t* pointerToSize = reinterpret_cast<std::size_t*>(newMem);
+        *pointerToSize = size;
+        return reinterpret_cast<void*>(newMem + offset);
+    }
+    else
+    {
+        return Z_NULL;
+    }
+}
+
+void aws_zfree(void * /* opaque */, void * ptr)
+{
+    if(ptr)
```

Comment:
ptr && ptr > offset

Reply:
that offset can be only 1, 4 or maybe 8. That is a way very little number to be valid memory anyway.